### PR TITLE
Reduce dependencies

### LINKED ISSUE #729: Reduce dependency usage
Some dependencies currently required by the project are marginally used and could potentially be removed. This acquires more relevance with the upcoming merge with k6 core due to the importance that is placed in this project to minimize the dependencies usage and limit its heterogeneity in order to prevent malfunctions between extensions code.

Some examples for dependencies that could potentially be reevaluated are:
- [github.com/hashicorp/go-multierror](https://github.com/grafana/xk6-browser/blob/5bd4fa1d7d9c9345eeee8107a97c021b7fee08b2/common/remote_object.go#L16)
- [github.com/fatih/color](https://github.com/grafana/xk6-browser/blob/5bd4fa1d7d9c9345eeee8107a97c021b7fee08b2/log/logger.go)
- ~~[github.com/mccutchen/go-httpbin/httpbin](https://github.com/grafana/xk6-browser/blob/5bd4fa1d7d9c9345eeee8107a97c021b7fee08b2/tests/ws/server.go#L22): Although in this case this one has 0 transitive dependencies~~ -> Not included in binary as is only required in `tests` package.
- [github.com/oxtoacart/bpool](https://github.com/grafana/xk6-browser/blob/5bd4fa1d7d9c9345eeee8107a97c021b7fee08b2/k6ext/k6test/vu.go#L15): Only 1 transitive dependency for testify, which is nevertheless used in xk6-browser anyway.

Other dependencies such as:
- golang.org/x/exp
- golang.org/x/net
- ~~golang.org/x/sync~~ -> Only used in `tests` pkg
- gopkg.in/guregu/null.v3
Are marginally used but are required as transitive dependencies by other modules.

Additionally this is in interest of reducing the xk6-browser / k6 binary size.

### PR DESCRIPTION
**Pull request by ka3de**
This PR tries to get rid of some dependencies currently required by the main module that are marginally used. More specifically:
- Removes the usage of `github.com/fatih/color`
- Removes the requirement for `golang.org/x/exp` from main module (still required as transitive)
  This produces the direct inclusion of `github.com/google/go-cmp` but that is only used inside `tests` pkg
- Replaces the usage of `github.com/hashicorp/go-multierror` with a custom implementation that fulfills our current use case

Closes #729.

### CODE REVIEW

**Review by ankur22 — CHANGES_REQUESTED** [review]
Thanks for doing this! Most of it looks good. I've left some suggestions that I think are worth considering, and let me know in the comments if my suggestions don't work 🙂 

EDIT: Could you also double check the CI pipeline issues and resolve those? There are some tests that fail often in the CI pipeline (we have raised issues for those so you should be able to find them, and then rerun the failed tests runs).

File: common/remote_object.go
Comment (ka3de) [reply common/remote_object.go]: Makes sense, fixed in [b4f3aad](https://github.com/grafana/xk6-browser/pull/734/commits/b4f3aad5ab0efc980a1fa40affe0dc0a32aafdce). Looks better now!

File: go.mod
Comment (ankur22) [inline go.mod:17]: Let's match the versions with k6 deps. For example, k6 is using `gopkg.in/guregu/null.v3 v3.3.0`

File: common/remote_object.go
Comment (ka3de) [reply common/remote_object.go]: The problem with `As()` AFAIK is that it will return the first error in the chain that matches the given type, if found any. That means that you might skip some "wrapped errors" in the chain, but considering the current implementation I think it should be safe here, changed [here](https://github.com/grafana/xk6-browser/blob/d76ff69916458824d567f80c8849a648b4efde8a/common/remote_object.go#L193).

File: common/remote_object.go
Comment (ka3de) [reply common/remote_object.go:42]: Absolutely. Done [here](https://github.com/grafana/xk6-browser/blob/d76ff69916458824d567f80c8849a648b4efde8a/common/remote_object_test.go#L240).

File: common/remote_object.go
Comment (ka3de) [reply common/remote_object.go]: I have refactored this function, take a look [here](https://github.com/grafana/xk6-browser/blob/d76ff69916458824d567f80c8849a648b4efde8a/common/remote_object.go#L67).

File: go.mod
Comment (ka3de) [reply go.mod]: Thanks for taking a look at that! Changed in [364f771](https://github.com/grafana/xk6-browser/pull/734/commits/364f7716220ecf3c36ac4a25be65eaf6fb0e8aa3).

File: log/logger.go
Comment (ankur22) [inline log/logger.go:90]: Let's remove this completely -- `l.Logger` cannot be `nil` at this point, since the actions before it would have failed will `nil` pointer errors.

File: common/remote_object.go
Comment (ankur22) [inline common/remote_object.go:50]: Can you also add a test for this `Error` function?

File: common/remote_object.go
Comment (ankur22) [reply common/remote_object.go]: yep, makes sense. maybe add a `nolint` directive here and explain why.

File: log/logger.go
Comment (ka3de) [reply log/logger.go:90]: I think the proper action here would be to, in the constructor [here](https://github.com/grafana/xk6-browser/blob/364f7716220ecf3c36ac4a25be65eaf6fb0e8aa3/log/logger.go#L35), provide a "default" logrus Logger in case the input one is `nil`. And then we can remove this part of the code. What do you think?

Comment (ka3de) [reply log/logger.go]: Done in: [61609ee](https://github.com/grafana/xk6-browser/pull/734/commits/61609ee2ccd55f921b99b83ca59228cdadd64d81).

File: common/remote_object.go
Comment (ka3de) [reply common/remote_object.go]: As mentioned, for our current implementation, using `As()` should be safe, changed.

File: common/remote_object.go
Comment (ka3de) [reply common/remote_object.go:50]: Sure, done in [c6e7e9a](https://github.com/grafana/xk6-browser/pull/734/commits/c6e7e9ac8cb3e3b610ffbb2905d8b88c4c444494).

File: log/logger.go
Comment (inancgumus) [reply log/logger.go]: @ka3de doing so makes more sense 👍

File: common/remote_object_test.go
Comment (inancgumus) [inline common/remote_object_test.go:275]: No need to say `.Error` as `Sprintf` will automatically expand an error to `string`.

```suggestion
			expStr: fmt.Sprintf(
				"3 errors occurred:\n\t* %s\n\t* %s\n\t* %s\n",
				mockErr1, mockErr2, mockErr3,
			),
```

File: common/remote_object_test.go
Comment (inancgumus) [inline common/remote_object_test.go:284]: Same here.

```suggestion
			expStr: fmt.Sprintf(
				"2 errors occurred:\n\t* %s\n\t* %s\n",
				mockErr1, mockErr3,
			),
```

File: common/remote_object_test.go
Comment (inancgumus) [inline common/remote_object_test.go:293]: Here too:

```suggestion
			expStr: fmt.Sprintf(
				"3 errors occurred:\n\t* %s\n\t* %s\n\t* %s\n",
				mockWrappedMultiErr, mockErr3, mockErr2,
			),
```

**Review by inancgumus — CHANGES_REQUESTED** [review]
Didn't look into this thoroughly yet, here is what I saw after skimming it.

It's great that we're getting rid of that excess calories 👏

File: common/remote_object_test.go
Comment (ka3de) [reply common/remote_object_test.go]: Fixed in [805bf89](https://github.com/grafana/xk6-browser/pull/734/commits/805bf899fd3e5cc0dac0221731416bf375273812).

File: common/remote_object_test.go
Comment (ka3de) [reply common/remote_object_test.go]: Fixed in [805bf89](https://github.com/grafana/xk6-browser/pull/734/commits/805bf899fd3e5cc0dac0221731416bf375273812).

File: log/logger.go
Comment (ankur22) [reply log/logger.go]: Could we log a warning too? just so that we're aware that we've forgotten to do something?

**Review by ankur22 — APPROVED** [review]
LGTM. Thanks for taking on the feedback 🙂

File: common/remote_object.go
Comment (inancgumus) [inline common/remote_object.go:42]: Would it make more sense to move this to `multierror.go` file? Instead of overcrowding this file or the already crowded `helpers.go` file. Also, IMO, having a test for this in `remote_object_test.go` file does not make sense.

File: log/logger.go
Comment (ka3de) [reply log/logger.go]: I think that makes sense! although here I see 2 options:
Comply with our logging API to produce the log:
```
// New creates a new logger.
func New(logger *logrus.Logger, iterID string) *Logger {
	var defLogger bool
	if logger == nil {
		defLogger = true
		logger = logrus.New()
	}
	l := &Logger{
		Logger: logger,
		iterID: iterID,
	}
	if defLogger {
		l.Warnf("Logger", "no logger supplied, using default")
	}

	return l
}
```
Or log directly with default logrus logger:
```
// New creates a new logger.
func New(logger *logrus.Logger, iterID string) *Logger {
	if logger == nil {
		logger = logrus.New()
		logger.Warn("no logger supplied, using default")
	}
	return &Logger{
		Logger: logger,
		iterID: iterID,
	}
}
```

What do you think?

Comment (ankur22) [reply log/logger.go]: My vote would be 1, but i'd be happy with either.

File: common/remote_object.go
Comment (inancgumus) [inline common/remote_object.go:62]: You could use `fmt.Fprintf` here instead:

```suggestion
	fmt.Fprintf(&buf, "%d errors occurred:\n", len(me.Errors))
	for _, e := range me.Errors {
		fmt.Fprintf(&buf, "\t* %s\n", e)
	}
```

**Review by inancgumus — CHANGES_REQUESTED** [review]
LGTM with some suggestions.

File: log/logger.go
Comment (inancgumus) [reply log/logger.go]: My vote is 2 since it's simple—1 is complicated.

> Debugging is twice as hard as writing the code in the first place. Therefore, if you write the code as cleverly as possible, you are, by definition, not smart enough to debug it." - Brian W. Kernighan

File: common/remote_object.go
Comment (ankur22) [reply common/remote_object.go:42]: I was thinking the same, but then I realised that `multierror` is only used in one file, so it makes sense to leave the type close to where it's used.

File: common/remote_object.go
Comment (ankur22) [reply common/remote_object.go:62]: Out of curiosity, why would you chose one over the other?

File: common/remote_object.go
Comment (inancgumus) [reply common/remote_object.go:42]: Yes, I thought that too, but as a helper, it still makes more sense to me in a new file. Also, in Go 1.20+, we will remove this file as there are multiple errors support is on the way. Having it in a new file will also make us spot it more easily.

File: log/logger.go
Comment (ka3de) [reply log/logger.go]: Ops, just pushed option 1 as I thought it was better to comply with our log output, but we can discuss!
The log message itself in this case I don't think it's a big deal, but at the same time I don't think the implementation for the first option is so cumbersome.

File: common/remote_object.go
Comment (ka3de) [reply common/remote_object.go:42]: Exactly the multierror implementation is only used in this file, and I would keep its definition inside this file only and try to not widespread its usage. Actually after the merge is done I would create an issue in order to get rid of this implementation or replace it by something simpler.

File: log/logger.go
Comment (inancgumus) [reply log/logger.go]: Less code is better than more code. But, 🤷 , I leave it to you. I said what I wanted to say :)

File: common/remote_object.go
Comment (inancgumus) [reply common/remote_object.go:62]: Yes, it's more performant and less code (=easier to read and maintain). More function calls often result in a less performant code.

File: common/remote_object.go
Comment (ka3de) [reply common/remote_object.go:42]: My reason to moving it to a new file would be if we think that there is a valid use case for "multierrors" and therefore want to use them in more places. In that case I think it's good to move it as a helper and improve its implementation in future releases. Otherwise I would keep it in this file and try to remove it for something simpler.

What do you think?

Comment (inancgumus) [reply common/remote_object.go:42]: @ka3de I would rather create that issue soon without waiting for the merge.

Comment (ka3de) [reply common/remote_object.go:42]: Done! See #738 .

Comment (inancgumus) [reply common/remote_object.go:42]: Thanks!

File: common/remote_object.go
Comment (ka3de) [reply common/remote_object.go]: You are right! Done.

**Review by inancgumus — APPROVED** [review]
🚀

**Pull request merged**
